pennylane/transforms/batch_transform.py

```diff
@@ -0,0 +1,239 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains tools and decorators for registering qfunc transforms."""
+# pylint: disable=too-few-public-methods
+import functools
+
+import pennylane as qml
+
+from pennylane.interfaces.batch import execute
+
+
+def _create_qnode_internal_wrapper(qnode, transform, targs, tkwargs):
+    def _wrapper(*args, **kwargs):
+        qnode.construct(args, kwargs)
+        tapes, processing_fn = transform.construct(qnode.qtape, *targs, **tkwargs)
+
+        # TODO: work out what to do for backprop
+        interface = qnode.interface
+
+        # TODO: extract gradient_fn from QNode
+        gradient_fn = qnode.diff_method
+
+        if interface is None or not transform.differentiable:
+            gradient_fn = None
+
+        elif gradient_fn in ("best", "parameter-shift"):
+            gradient_fn = qml.gradients.param_shift
+
+        elif gradient_fn == "finite-diff":
+            gradient_fn = qml.gradients.finite_diff
+
+        res = execute(
+            tapes,
+            device=qnode.device,
+            gradient_fn=gradient_fn,
+            interface=interface,
+        )
+
+        return processing_fn(res)
+
+    return _wrapper
+
+
+class batch_transform:
+    r"""Class for registering a tape transform that takes a tape, and outputs
+    a batch of tapes to be independently executed on a quantum device.
+
+    Examples of such transforms include quantum gradient shift rules (such
+    as finite-differences and the parameter-shift rule) and metrics such as
+    the quantum Fisher information matrix.
+
+    Args:
+        transform_fn (function): The function to register as the batch tape transform.
+            It can have an arbitrary number of arguments, but the first argument
+            **must** be the input tape.
+        expand_fn (function): An expansion function (if required) to be applied to the
+            input tape before the transformation takes place.
+        differentiable (bool): Specifies whether the transform is differentiable or
+            not.
+
+    **Example**
+
+    A valid batch tape transform is a quantum function that satisfies the following:

```

Comment:
```suggestion
    A valid batch tape transform is a function that satisfies the following:
```
(It's not a quantum function?)